### ISSUE TITLE
Refactoring Ideas

### ISSUE THREAD

**Issue by scttl** [NervanaSystems/neon#62]
We're starting to embark upon a fairly substantial refactoring of the neon codebase in an effort to make it easier to use, and clean out some of the accumulated cruft.

We'd like to use this issue to get feedback from the broader community, and give a heads up that changes are on their way.  What do you like, not like?  Some of the things we're initially considering include:

* Generic global backend and tensor with op-tree support, reusable buffers
* reworking Layers to separate out state, support deferred initialization
* Dataset revamping.  Simplify loading, reading tabular data, DataWorker integration
* Replace YAML with JSON?  Make it easier to dump model specifications back out from loaded models

**Comment by seba-1511** [comment]
One thing I believe works fairly well is the `gen_backend()` method, because it allows to easily switch from one backend to another. Adding a way to specify how the backend should be distributed (how many processes, etc...) programatically - as opposed to doing that when executing MPI - would be cool, but might not fit as a refactoring issue.

In addition, I like the idea of specifying models with JSON. It would be really nice to specify a model in JSON and then do something like `load_model('path_to_json')` and get back the model. I recently found out about util.persist.deserialize() but haven't used it yet. `load_model` could take parameters that would overwrite parameters in the JSON.

Finally, as you pointed out, working with datasets is a bit of a mess. Especially, it is strange that doing stuff such as pre-processing the data involves writing your own Dataset class. And having the scikit-learn interface is a must have.

**Comment by pranv** [comment]
:+1: for JSON

**Comment by seba-1511** [comment]
One additional thing that I find difficult to do and that might be useful, is to easily convert the backend and some parameters of the models. Example, I train a MLP on 4 GPUs with batch size 128 on Imagenet, and then I want to run it on a Raspberry Pi (on CPU) with batch size 1. Again, not sure if it fits in the refactoring.

**Comment by llllaaaa** [comment]
I think automatic differentiation may be introduced with the refactoring, then the activation functions and cost functions can be written easily without providing the derivative function.

**Comment by pranv** [comment]
This may be a bit out order, but here it is:

Why can't neon have a Keras style API?

**Comment by llllaaaa** [comment]
I have some suggestions on RNN

1. Allow the output of t0 to be the input of t1
2. Allow saving and setting the states of the network, then the trained network can move forward step by step. This will be very useful in real time application. 
3. Support variations of RNN, such as GRU, Gated Feedback LSTM, Gated Feedback GRU, Gird LSTM

**Comment by tambetm** [comment]
I just wanted to add, that layer methods mentioned in #65 were confusing to me as well. I expected layers to have two simple methods: forward() and backward(). Instead tanh had three methods, all of which did forward pass first and two of them did backward pass as well. That seems like a lot of redundancy, although as I understand, apply_derivatives() (and possibly also apply_function()?) is not called at all. At least they would deserve clarifying comment.

BTW, I implemented tanh version optimized for normalized inputs, as described in ["Efficient BackProp"](http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf). This takes form of f(x) = 1.7159*tanh(x*2/3). At least in my case it improved results considerably. If anybody is interested, the code is here: https://github.com/tambetm/neon/commit/f31fc675a6f652bfcbeeda21ced0d581c4e14ff6.

**Comment by BenjaminBossan** [comment]
My biggest wish: Making it possible to skip the Dataset class and use numpy arrays as default data containers instead (such as in scikit learn). Despite the documentation, I found it painful to use neon with my own data.

Personally, I do not care for yaml or json model definitions. I find the model defintion in pure python very succinct and readable.

I'm eager to use neon more, keep up the great work.

**Comment by avostryakov** [comment]
1. I want to see LSTM support with mini-batches and multi-GPUs support. 
2. For NLP tasks, it's cool to have EmbeddingLayer with possibility to switch off training of embedding matrix like following in Lasagne: http://lasagne.readthedocs.org/en/latest/modules/layers/embedding.html#lasagne.layers.EmbeddingLayer
